backend/de.metas.contracts/src/main/java/de/metas/contracts/commission/salesrep/process/CreateMissingCommissionEntriesForSalesRep.java

```diff
@@ -0,0 +1,421 @@
+
+					valueByInterval.put(oldValueInterval, logEntry.getOldValue());
+				}
+
+				if (logEntry.getNewValue() != null)
+				{
+					final InstantInterval newValueInterval = InstantInterval.of(logEntry.getChangeTimestamp(), targetInterval.getTo());
+
+					valueByInterval.put(newValueInterval, logEntry.getNewValue());
+				}
+			}
+			else if (logEntry.getOldValue() != null )
+			{
+				final InstantInterval oldValueInterval = InstantInterval.of(currentFrom, logEntry.getChangeTimestamp());
+
+				valueByInterval.put(oldValueInterval, logEntry.getOldValue());
+			}
+
+			currentFrom = logEntry.getChangeTimestamp();
+		}
+
+		return ImmutableMap.copyOf(valueByInterval);
+	}
+
+	private Optional<BPartnerId> getSuperSalesRepBPId(@NonNull final String superSalesRepUserIdValue, @NonNull final BPartnerId salesRepId)
+	{
+		final int userId = NumberUtils.asInt(superSalesRepUserIdValue, -1);
+		final UserId superSalesRepUserId = UserId.ofRepoIdOrNull(userId);
+
+		if (superSalesRepUserId == null)
+		{
+			Loggables.withLogger(logger, Level.WARN)
+					.addLog("*** WARN during calculation for bPartnerId: {}:"
+									+ " salesRepUser value: {} cannot be converted to int! Skipping.. ", superSalesRepUserIdValue);
+			return Optional.empty();
+		}
+
+		final I_AD_User superSalesRepUser = userDAO.getById(superSalesRepUserId);
+
+		if (superSalesRepUser.getC_BPartner_ID() <= 0)
+		{
+			Loggables.withLogger(logger, Level.WARN)
+					.addLog("*** WARN during calculation for bPartnerId: {}:"
+									+ " salesRepUser: {} doesn't have a bPartner! Skipping.. ", salesRepId, superSalesRepUser.getAD_User_ID());
+			return Optional.empty();
+		}
+		else
+		{
+			final I_C_BPartner superSalesRepBPartner = bPartnerDAO.getById(superSalesRepUser.getC_BPartner_ID());
+
+			if (!superSalesRepBPartner.isSalesRep())
+			{
+				Loggables.withLogger(logger, Level.WARN)
+						.addLog("*** WARN during calculation for bPartnerId: {}: missing 'isSalesRep' flag on regional manager: {}! Skipping.. ",
+								salesRepId, superSalesRepBPartner.getC_BPartner_ID());
+
+				return Optional.empty();
+			}
+			else
+			{
+				return Optional.of(BPartnerId.ofRepoId(superSalesRepBPartner.getC_BPartner_ID()));
+			}
+		}
+	}
+
+	public void recalculateCommissionForCriteria(@NonNull final RecalculateCommissionCriteria commissionCriteria)
+	{
+		bPartnerBL.setBPartnerSalesRepId(commissionCriteria.getBPartnerId(), commissionCriteria.getTopLevelSalesRepId());
+
+		final InvoiceCandidateQuery invoiceCandidateQuery = InvoiceCandidateQuery.builder()
+				.dateToInvoiceInterval(commissionCriteria.getTargetInterval())
+				.salesRepBPartnerId(commissionCriteria.getBPartnerId())
+				.build();
+
+		final InvoiceCandidateMultiQuery multiQuery = InvoiceCandidateMultiQuery.builder().query(invoiceCandidateQuery).build();
+
+		final List<I_C_Invoice_Candidate> invoiceCandidates = invoiceCandDAO.getByQuery(multiQuery);
```

Comment:
we need to do the same for C_Invoice..we have aprox *200* commission instances which were triggered by an invoice or a credit memo